# 合併地震資料及爬文資料

In [1]:
import numpy as np
import pandas as pd

In [2]:
df_earthquake = pd.read_csv("earthquake_data_0620.csv")
df_ptt = pd.read_csv("ptt.csv")

In [3]:
df_earthquake.columns = ['編號', 'Date_format', 'Longitude', 'Latitude', '規模', '深度', '位置', '震度']

In [4]:
df_earthquake["IsOfficialReport"] = 1

In [5]:
df_earthquake.head()

編號          Date_format  Longitude  Latitude   規模    深度  \
0  小區域有感地震  2020-06-17 08:06:39     120.56     23.50  3.1   8.3   
1       33  2020-06-17 07:17:00     121.98     24.62  4.7  60.8   
2  小區域有感地震  2020-06-15 10:42:24     121.63     24.33  3.1  19.1   
3  小區域有感地震  2020-06-15 02:15:40     121.66     24.39  3.6  35.0   
4  小區域有感地震  2020-06-15 02:10:08     121.62     24.40  4.3  40.7   

                                       位置   震度  IsOfficialReport  
0    嘉義縣政府東方  28.0  公里 (位於嘉義縣竹崎鄉)(嘉義縣竹崎鄉)  3.0                 1  
1  宜蘭縣政府東偏南方  24.9  公里 (位於臺灣東部海域)(臺灣東部海域)  3.0                 1  
2    花蓮縣政府北方  37.6  公里 (位於花蓮縣秀林鄉)(花蓮縣秀林鄉)  1.0                 1  
3  宜蘭縣政府南偏西方  39.6  公里 (位於宜蘭縣南澳鄉)(宜蘭縣南澳鄉)  1.0                 1  
4  宜蘭縣政府南偏西方  39.3  公里 (位於宜蘭縣南澳鄉)(宜蘭縣南澳鄉)  2.0                 1

In [6]:
df_ptt["IsOfficialReport"] = 0

In [7]:
df_ptt.head()

Unnamed: 0                     Author      Board         City  \
0           0          werqq (sandy0980)  Gossiping       Taipei   
1           1                  inmee (J)  Gossiping       Taipei   
2           2  MiLuDaiBoom (ⓂⓛⓁⓤⒹⓐⓛⒷⓞⓞⓂ)  Gossiping       Taipei   
3           3               somehog (低進)  Gossiping       Taipei   
4           4                WeiU (微，Ｕ文)  Gossiping  Tainan City   

                                             Content  \
0  嗨 大家早安\n\n小妹剛剛被地震震了一下\n\n不曉得被震這一下後\n\n莎莎會不會受傷呢...   
1  剛剛上下震了好大一下\n\n讓小妹嚇得花枝亂戰\n\n一時之間不曉得該怎麼辦\n\n究竟.....   
2  \n\n\n欸欸欸\n\n剛剛在搖欸\n\n阿\n\n小妹我還在睡覺\n\n當然沒穿衣服啊\...   
3                    為什麼我又沒收到國家警報了?\n手機明明滿格!\n幹~\n\n   
4                       \n\n妹子我喜歡妳！！\n\n但地震了快逃啊！\n\n   

                       Date               Full IP              IP  \
0  Thu Jun 11 09:44:31 2020   223.136.205.58 (臺灣)  223.136.205.58   
1  Thu Jun 11 09:44:05 2020   223.137.176.72 (臺灣)  223.137.176.72   
2  Thu Jun 11 09:43:23 2020   110.50.129.111 (臺灣)  110.50.129.111   
3  Thu Jun 11 09:42:46 2020    119.14.24.108 (臺灣)   119.14.24.108   
4  Thu Jun 11 09:41:42 2020   119.14.206.234 (臺灣)  119.14.206.234   

  Longitude and Latitude                    Mobile  \
0   經度：121.532　緯度：25.048          Chunghwa Telecom   
1   經度：121.532　緯度：25.048   Taiwan Academic Network   
2   經度：121.532　緯度：25.048   Taiwan Academic Network   
3   經度：121.532　緯度：25.048  Hoshin Multimedia Center   
4   經度：120.213　緯度：22.991  Hoshin Multimedia Center   

                                Nation  \
0  AS - Asia ( 亞洲 ) TW - Taiwan ( 台灣 )   
1  AS - Asia ( 亞洲 ) TW - Taiwan ( 台灣 )   
2  AS - Asia ( 亞洲 ) TW - Taiwan ( 台灣 )   
3  AS - Asia ( 亞洲 ) TW - Taiwan ( 台灣 )   
4  AS - Asia ( 亞洲 ) TW - Taiwan ( 台灣 )   

                                                Push                  Title  \
0  ['', '', '本網站已依台灣網站內容分級規定處理。此區域為限制級，未滿十八歲者不得瀏覽...       [問卦] 地震後愛莉莎莎沒事吧？   
1  ['', '', '本網站已依台灣網站內容分級規定處理。此區域為限制級，未滿十八歲者不得瀏覽...  [問卦] 地震第一時間要做什麼才是正常人？   
2  ['', '', '本網站已依台灣網站內容分級規定處理。此區域為限制級，未滿十八歲者不得瀏覽...       [問卦] 地震時沒穿衣服怎麼辦？   
3  ['', '', '本網站已依台灣網站內容分級規定處理。此區域為限制級，未滿十八歲者不得瀏覽...    [爆卦] 幹!地震 為什麼我又邊緣人了   
4  ['', '', '本網站已依台灣網站內容分級規定處理。此區域為限制級，未滿十八歲者不得瀏覽...               [爆卦] 地震！   

   YouWin  Longitude  Latitude          Date_format  IsOfficialReport  
0       0    121.532    25.048  2020-06-11 09:44:31                 0  
1       0    121.532    25.048  2020-06-11 09:44:05                 0  
2       0    121.532    25.048  2020-06-11 09:43:23                 0  
3       0    121.532    25.048  2020-06-11 09:42:46                 0  
4       0    120.213    22.991  2020-06-11 09:41:42                 0

## 合併檔案

In [8]:
df = pd.merge(df_earthquake, df_ptt,how='outer', on = ['Date_format', 'Longitude', 'Latitude', 'IsOfficialReport'], sort="Date_format")

df["TimeDifference"] = "NaN"

In [9]:
df.head()

編號          Date_format  Longitude  Latitude  規模  深度   位置  震度  \
0  NaN  2008-05-16 11:31:40    121.000    23.500 NaN NaN  NaN NaN   
1  NaN  2009-07-14 19:18:33    121.000    23.500 NaN NaN  NaN NaN   
2  NaN  2009-08-13 22:04:59    121.000    23.500 NaN NaN  NaN NaN   
3  NaN  2010-03-18 22:54:47    120.684    24.147 NaN NaN  NaN NaN   
4  NaN  2011-03-12 02:22:52    139.690    35.690 NaN NaN  NaN NaN   

   IsOfficialReport  Unnamed: 0  ...    \
0                 0       998.0  ...     
1                 0       997.0  ...     
2                 0       996.0  ...     
3                 0       995.0  ...     
4                 0       994.0  ...     

                                             Content  \
0  ※ 引述《makihara (maki)》之銘言：\n: 看新聞很多截肢的人...而且很多是...   
1  轉自SHARE板\n\n 作者  mstar (Wayne Su)             ...   
2                                         這是國內一位河圖洛書   
3  ※ 引述《GundamXX》之銘言：\n\n\n劣退或水桶都好，五樓不要那麼沒品，要是有你要...   
4  \n下午那時剛要搭新宿中央縣回家,那時的我在地下大概1~2層的底下\n\n突然間輕微的搖晃,...   

                       Date                              Full IP  \
0  Fri May 16 11:31:40 2008                                  NaN   
1  Tue Jul 14 19:18:33 2009   61.229.29.48         (05/13 10:53)   
2  Thu Aug 13 22:04:59 2009   61.20.131.104        (08/13 22:53)   
3  Thu Mar 18 22:54:47 2010   112.104.159.100      (03/18 22:58)   
4  Sat Mar 12 02:22:52 2011   119.107.220.15       (03/12 02:25)   

                IP Longitude and Latitude          Mobile  \
0     61.229.29.48   經度：121.000　緯度：23.500           HiNet   
1     61.229.29.48   經度：121.000　緯度：23.500           HiNet   
2    61.20.131.104   經度：121.000　緯度：23.500      FarEasTone   
3  112.104.159.100   經度：120.684　緯度：24.147  Digital United   
4   119.107.220.15   經度：139.690　緯度：35.690      au one net   

                                Nation     Push  \
0  AS - Asia ( 亞洲 ) TW - Taiwan ( 台灣 )  No Push   
1  AS - Asia ( 亞洲 ) TW - Taiwan ( 台灣 )  No Push   
2  AS - Asia ( 亞洲 ) TW - Taiwan ( 台灣 )  No Push   
3  AS - Asia ( 亞洲 ) TW - Taiwan ( 台灣 )  No Push   
4   AS - Asia ( 亞洲 ) JP - Japan ( 日本 )  No Push   

                            Title YouWin  
0      Re: [新聞] 四川地震「別鋸我的腿 我寧願自殺」    0.0  
1   Re: [問卦] 有沒有地震震度都用幾顆原子彈來表示的 …    0.0  
2  [爆卦] 為什麼87水災921地震88水災全部在尾數八的年?    0.0  
3         Re: [問卦] 有沒有明天早上會大地震的八卦    0.0  
4                  [爆卦] 大地震時在新宿的我    0.0  

[5 rows x 23 columns]

In [10]:
df.tail()

編號          Date_format  Longitude  Latitude  規模  深度   位置  震度  \
7323  NaN  2020-06-17 08:51:57    121.742    25.128 NaN NaN  NaN NaN   
7324  NaN                  NaN    121.532    25.048 NaN NaN  NaN NaN   
7325  NaN                  NaN    121.532    25.048 NaN NaN  NaN NaN   
7326  NaN                  NaN    121.532    25.048 NaN NaN  NaN NaN   
7327  NaN                  NaN    121.612    23.981 NaN NaN  NaN NaN   

      IsOfficialReport  Unnamed: 0  ...    \
7323                 0      1821.0  ...     
7324                 0       541.0  ...     
7325                 0       731.0  ...     
7326                 0      1453.0  ...     
7327                 0       763.0  ...     

                                                Content  \
7323  \n\n欸欸欸欸欸\n\n最近覺得地震特別搖\n\n但看新聞都說還好\n\n所以是我太敏感\...   
7324                         台南地震\n\n東區有感\n\n不過沒有很大\n\n   
7325  時間: Thu Aug  8 05:28:26 2019\n\n幹 超大\n\n\n地震\n...   
7326  1.媒體來源:\n自由時報\n\n\n2.記者署名\n曾德蓉\n\n\n3.完整新聞標題:\...   
7327  \n第二次地震，怕怕的\n等等補資訊\nhttp://i.imgur.com/ZSd8NlT...   

                          Date               Full IP              IP  \
7323  Wed Jun 17 08:51:57 2020   110.50.130.167 (臺灣)  110.50.130.167   
7324   Mon Sep 30 18:39:32 201   115.82.140.129 (臺灣)  115.82.140.129   
7325                EARTHquake   223.137.160.14 (臺灣)  223.137.160.14   
7326                       NaN                   NaN     39.12.35.52   
7327                       NaN     59.115.16.25 (臺灣)    59.115.16.25   

     Longitude and Latitude            Mobile  \
7323   經度：121.742　緯度：25.128      Vibo Telecom   
7324   經度：121.532　緯度：25.048     Taiwan Mobile   
7325   經度：121.532　緯度：25.048  Chunghwa Telecom   
7326   經度：121.532　緯度：25.048        FarEasTone   
7327   經度：121.612　緯度：23.981             HiNet   

                                   Nation  \
7323  AS - Asia ( 亞洲 ) TW - Taiwan ( 台灣 )   
7324  AS - Asia ( 亞洲 ) TW - Taiwan ( 台灣 )   
7325  AS - Asia ( 亞洲 ) TW - Taiwan ( 台灣 )   
7326  AS - Asia ( 亞洲 ) TW - Taiwan ( 台灣 )   
7327  AS - Asia ( 亞洲 ) TW - Taiwan ( 台灣 )   

                                                   Push  \
7323  ['→ Usayhiisayha: 跟跌倒的人體重有關 06/17 08:52', '推 V...   
7324  ['噓 r30385:36.229.226.172 09/30 18:39', '→ joh...   
7325  ['推 Luciferspear: 超大  114.24.225.24 08/08 05:2...   
7326                                            No Push   
7327  ['噓 r30385:42.73.78.14 08/06 09:32', '→ Seles:...   

                         Title YouWin  
7323        [問卦] 地震搖晃程度和屋齡有關嗎？    0.0  
7324                   [爆卦] 地震    1.0  
7325                 Gossiping    1.0  
7326                       NaN    0.0  
7327  Tue Aug  6 09:32:01 2019    1.0  

[5 rows x 23 columns]

# Cleaning

In [11]:
df.iloc[100]["IsOfficialReport"]==1

True

In [12]:
earthquake_index = []
for i in range(len(df)):
    if df.iloc[i]["IsOfficialReport"]==1:
        if df.iloc[i+1]["YouWin"]==1: #在一起地震後確實有「你贏了」推文的那些文章 i.e. 地震後有地震文
            #df_new = df_new.append(df.iloc[i], ignore_index=True)
            earthquake_index.append(i)

In [13]:
len(earthquake_index)

91

代表有91起地震是在其後有「你贏了」推文的

In [14]:
df.iloc[earthquake_index].head()

編號          Date_format  Longitude  Latitude   規模     深度  \
4959  小區域有感地震  2019-05-08 23:48:47     122.49     24.82  5.1  122.7   
4965  小區域有感地震  2019-05-10 21:49:30     121.92     24.50  3.9   40.4   
4968       36  2019-05-13 16:55:21     121.90     24.75  5.2   88.0   
5013  小區域有感地震  2019-05-19 00:38:34     122.19     24.03  4.7   22.5   
5025       37  2019-05-22 09:37:24     121.64     23.96  4.8   28.8   

                                        位置   震度  IsOfficialReport  Unnamed: 0  \
4959  宜蘭縣政府東方  74.2  公里 (位於臺灣東部海域)(臺灣東部海域)  2.0                 1         NaN   
4965   宜蘭縣政府東南方  30.3  公里 (位於宜蘭縣近海)(宜蘭縣近海)  3.0                 1         NaN   
4968    宜蘭縣政府東方  14.1  公里 (位於宜蘭縣近海)(宜蘭縣近海)  3.0                 1         NaN   
5013  花蓮縣政府東方  58.3  公里 (位於臺灣東部海域)(臺灣東部海域)  1.0                 1         NaN   
5025    花蓮縣政府東南方  4.0  公里 (位於花蓮縣近海)(花蓮縣近海)  4.0                 1         NaN   

      ...   Content Date Full IP   IP Longitude and Latitude Mobile Nation  \
4959  ...       NaN  NaN     NaN  NaN                    NaN    NaN    NaN   
4965  ...       NaN  NaN     NaN  NaN                    NaN    NaN    NaN   
4968  ...       NaN  NaN     NaN  NaN                    NaN    NaN    NaN   
5013  ...       NaN  NaN     NaN  NaN                    NaN    NaN    NaN   
5025  ...       NaN  NaN     NaN  NaN                    NaN    NaN    NaN   

     Push Title YouWin  
4959  NaN   NaN    NaN  
4965  NaN   NaN    NaN  
4968  NaN   NaN    NaN  
5013  NaN   NaN    NaN  
5025  NaN   NaN    NaN  

[5 rows x 23 columns]

### 進一步判斷一起地震其後有多少「地震文」

In [15]:
earthquake_index

[4959,
 4965,
 4968,
 5013,
 5025,
 5030,
 5046,
 5051,
 5065,
 5076,
 5186,
 5197,
 5199,
 5232,
 5248,
 5262,
 5266,
 5278,
 5305,
 5315,
 5319,
 5324,
 5361,
 5490,
 5508,
 5511,
 5538,
 5566,
 5605,
 5630,
 5644,
 5649,
 5655,
 5682,
 5704,
 5718,
 5731,
 5739,
 5758,
 5763,
 5769,
 5774,
 5792,
 5795,
 5802,
 5807,
 5809,
 5866,
 5874,
 5895,
 5901,
 5994,
 6035,
 6039,
 6064,
 6072,
 6078,
 6107,
 6119,
 6126,
 6176,
 6192,
 6232,
 6289,
 6291,
 6340,
 6367,
 6369,
 6400,
 6406,
 6464,
 6498,
 6519,
 6530,
 6563,
 6569,
 6619,
 6785,
 6797,
 6919,
 6927,
 6972,
 7018,
 7067,
 7095,
 7166,
 7174,
 7192,
 7239,
 7248,
 7311]

In [16]:
df_new = pd.DataFrame()
for i in earthquake_index:
    j = i #indicator for appending 地震文
    df_new = df_new.append(df.iloc[j], ignore_index=True)
    #試著把震央位置加進來
    #df_new.iloc[len(df_new)-1]["CenterLocation"] = df_new.iloc[i]["位置"]
    j+=1
    while df.iloc[j]["IsOfficialReport"]!=1: #地震報告後東西是貼文的話就要append進來
        df_new = df_new.append(df.iloc[j], ignore_index=True)
        #試著把震央位置加進來
        #df_new.iloc[len(df_new)-1]["CenterLocation"] = df_new.iloc[i]["位置"]
        j+=1 # keep appending 地震文 till 遇到offical report
len(df_new)

723

## 重拿一次氣象局地震報告的index

In [17]:
earthquake_index = []
for i in range(len(df_new)):
    if df_new.iloc[i]["IsOfficialReport"]==1:
        earthquake_index.append(i)
print(earthquake_index)

[0, 4, 7, 35, 44, 46, 52, 56, 64, 73, 75, 79, 81, 85, 91, 93, 96, 99, 104, 107, 110, 112, 143, 166, 180, 182, 184, 186, 191, 195, 199, 202, 204, 212, 234, 237, 240, 247, 250, 255, 260, 265, 268, 270, 276, 278, 280, 288, 293, 297, 299, 312, 318, 322, 324, 332, 338, 345, 357, 364, 397, 411, 440, 469, 471, 481, 488, 490, 493, 499, 505, 510, 515, 526, 530, 533, 540, 553, 556, 559, 561, 563, 577, 584, 593, 600, 604, 607, 654, 663, 712]


In [18]:
str(df_new.iloc[1]['位置'])

'nan'

In [19]:
CenterLoc = []
for i in earthquake_index:
    j = i
    #if str(df_new.iloc[j]["位置"]) != "nan":
    CenterLoc.append(df_new.iloc[j]["位置"])
    j+=1
    try:
        while str(df_new.iloc[j]["位置"]) == "nan":
            CenterLoc.append(df_new.iloc[i]["位置"]) #如果是地震文，就把震央位置給填上 地震報告 的位置
            
            j+=1
    except:
        pass
len(CenterLoc)

723

In [20]:
# 同時處理震央及震度
CenterLoc = []
ShakingExtent = []
for i in earthquake_index:
    j = i
    CenterLoc.append(df_new.iloc[j]["位置"])
    ShakingExtent.append(df_new.iloc[i]["震度"])
    j+=1
    try:
        while str(df_new.iloc[j]["位置"]) == "nan":
            CenterLoc.append(df_new.iloc[i]["位置"]) #如果是地震文，就把震央位置給填上 地震報告 的位置
            ShakingExtent.append(df_new.iloc[i]["震度"])
            j+=1
    except:
        pass
len(CenterLoc)
len(ShakingExtent)

723

In [21]:
CenterLoc[0:10]

['宜蘭縣政府東方  74.2  公里 (位於臺灣東部海域)(臺灣東部海域)',
 '宜蘭縣政府東方  74.2  公里 (位於臺灣東部海域)(臺灣東部海域)',
 '宜蘭縣政府東方  74.2  公里 (位於臺灣東部海域)(臺灣東部海域)',
 '宜蘭縣政府東方  74.2  公里 (位於臺灣東部海域)(臺灣東部海域)',
 '宜蘭縣政府東南方  30.3  公里 (位於宜蘭縣近海)(宜蘭縣近海)',
 '宜蘭縣政府東南方  30.3  公里 (位於宜蘭縣近海)(宜蘭縣近海)',
 '宜蘭縣政府東南方  30.3  公里 (位於宜蘭縣近海)(宜蘭縣近海)',
 '宜蘭縣政府東方  14.1  公里 (位於宜蘭縣近海)(宜蘭縣近海)',
 '宜蘭縣政府東方  14.1  公里 (位於宜蘭縣近海)(宜蘭縣近海)',
 '宜蘭縣政府東方  14.1  公里 (位於宜蘭縣近海)(宜蘭縣近海)']

In [22]:
ShakingExtent[0:10]

[2.0, 2.0, 2.0, 2.0, 3.0, 3.0, 3.0, 3.0, 3.0, 3.0]

In [23]:
df_new["CenterLoc"] = CenterLoc
df_new["ShakingExtent"] = ShakingExtent

In [24]:
df_new

Author      Board              City  \
0                       NaN        NaN               NaN   
1       DevNull (/dev/null)  Gossiping     Mountain View   
2         jackyuki1 (喵喵快來 )  Gossiping            Taipei   
3             Dihang (阿美阿美)  Gossiping            Taipei   
4                       NaN        NaN               NaN   
5       tony77998 (偷尼柒柒玖玖捌)  Gossiping            Taipei   
6      chamge7788 (經濟王7887)  Gossiping  Taoyuan District   
7                       NaN        NaN               NaN   
8                  ros (螺絲)  Gossiping            Taipei   
9      samok (記者只會抄公社低卡記錄7)  Gossiping            Taipei   
10   world210 (KillerQueen)  Gossiping            Taipei   
11    LiaMeow (有時候甜蜜的心花朵朵開)  Gossiping            Taipei   
12                 a382 (斐)  Gossiping            Taipei   
13            cookie0215 ()  Gossiping            Taipei   
14    gcobs0834 (gcobs0834)  Gossiping            Taipei   
15       freelysoul (帥氣小活寶)  Gossiping            Taipei   
16   twin1949tw (台北Station)  Gossiping            Taipei   
17              qx5 (學豬裝可愛)  Gossiping           Hsinchu   
18            harry68a (結衣)  Gossiping            Taipei   
19         a055035 (tomato)  Gossiping          Taichung   
20        longhandsome (阿駥)  Gossiping               NaN   
21       HayamaKami (台大彭于晏)  Gossiping            Taipei   
22              JEWH (小田切腹)  Gossiping          Taichung   
23         airwalk66996 (平)  Gossiping            Taipei   
24          kevin600339 ( )  Gossiping            Taipei   
25            a386036 (骷髏怪)  Gossiping            Taipei   
26       chunyu0519 (廣東粥杰倫)  Gossiping         Chang-hua   
27                ddt258 ()  Gossiping            Taipei   
28        jasontzymann (梵天)  Gossiping            Taipei   
29              IHVH (讓愛滋生)  Gossiping       Nantou City   
..                      ...        ...               ...   
693     ButterflyBao (一面湖水)  Gossiping            Taipei   
694         aa9510883 (阿狄男)  Gossiping            Taipei   
695         aa9510883 (阿狄男)  Gossiping            Taipei   
696      e1608 (silentness)  Gossiping             Tampa   
697      e1608 (silentness)  Gossiping            Taipei   
698              smd1201 ()  Gossiping            Taipei   
699              smd1201 ()  Gossiping               NaN   
700        yutapon (川島  永嗣)  Gossiping            Taipei   
701        yutapon (川島  永嗣)  Gossiping               NaN   
702       purine5566 (漂亮56)  Gossiping            Taipei   
703       purine5566 (漂亮56)  Gossiping            Taipei   
704        wuyiulin (機掰客家人)  Gossiping          Taichung   
705        wuyiulin (機掰客家人)  Gossiping            Taipei   
706         Giornno (快樂吳郭魚)  Gossiping            Taipei   
707         Giornno (快樂吳郭魚)  Gossiping            Taipei   
708              xxxqma (昀)  Gossiping            Taipei   
709              HMTking ()  Gossiping            Taipei   
710              HMTking ()  Gossiping            Taipei   
711            IkariD (怡卡滴)  Gossiping            Taipei   
712                     NaN        NaN               NaN   
713      spirit119 (一二三木頭人)  Gossiping            Taipei   
714            meimeiamy ()  Gossiping            Taipei   
715   Ccsleep (notsleepyet)  Gossiping            Taipei   
716            a435007 (陣雨)  Gossiping            Taipei   
717            sam45 (流水年華)  Gossiping          Taichung   
718               star07 ()  Gossiping          Taichung   
719   open20023 (open20023)  Gossiping            Taipei   
720   air121 (I GOT SPIRIT)  Gossiping            Taipei   
721        musashi4735 (武藏)  Gossiping            Taipei   
722           king1010 (玩酷)  Gossiping            Taipei   

                                               Content  \
0                                                  NaN   
1    發震時間：2019/05/08 23:48:47\n位置：北緯 24.81度，東經 122....   
2    為什麼每次地震完\n\n小弟想上ptt 都上不去\n\n大家手腳都好快....馬上就上ptt...   
3    我國東北的三寶是什麼？\n\n人參，貂皮，烏拉草\n\n還是\n\n人參，貂皮，鹿茸\n\n...   
4         

In [25]:
df_new = df_new[['Author', 'City', 'Content', 'Date_format', 'Full IP',
       'IP', 'IsOfficialReport', 'Latitude', 'Longitude',
        'Mobile', 'Nation', 'Push',
       'Title', 'YouWin', '位置', '深度', '編號', '規模', '震度', 'CenterLoc', 'ShakingExtent']]

In [26]:
df_new

Author              City  \
0                       NaN               NaN   
1       DevNull (/dev/null)     Mountain View   
2         jackyuki1 (喵喵快來 )            Taipei   
3             Dihang (阿美阿美)            Taipei   
4                       NaN               NaN   
5       tony77998 (偷尼柒柒玖玖捌)            Taipei   
6      chamge7788 (經濟王7887)  Taoyuan District   
7                       NaN               NaN   
8                  ros (螺絲)            Taipei   
9      samok (記者只會抄公社低卡記錄7)            Taipei   
10   world210 (KillerQueen)            Taipei   
11    LiaMeow (有時候甜蜜的心花朵朵開)            Taipei   
12                 a382 (斐)            Taipei   
13            cookie0215 ()            Taipei   
14    gcobs0834 (gcobs0834)            Taipei   
15       freelysoul (帥氣小活寶)            Taipei   
16   twin1949tw (台北Station)            Taipei   
17              qx5 (學豬裝可愛)           Hsinchu   
18            harry68a (結衣)            Taipei   
19         a055035 (tomato)          Taichung   
20        longhandsome (阿駥)               NaN   
21       HayamaKami (台大彭于晏)            Taipei   
22              JEWH (小田切腹)          Taichung   
23         airwalk66996 (平)            Taipei   
24          kevin600339 ( )            Taipei   
25            a386036 (骷髏怪)            Taipei   
26       chunyu0519 (廣東粥杰倫)         Chang-hua   
27                ddt258 ()            Taipei   
28        jasontzymann (梵天)            Taipei   
29              IHVH (讓愛滋生)       Nantou City   
..                      ...               ...   
693     ButterflyBao (一面湖水)            Taipei   
694         aa9510883 (阿狄男)            Taipei   
695         aa9510883 (阿狄男)            Taipei   
696      e1608 (silentness)             Tampa   
697      e1608 (silentness)            Taipei   
698              smd1201 ()            Taipei   
699              smd1201 ()               NaN   
700        yutapon (川島  永嗣)            Taipei   
701        yutapon (川島  永嗣)               NaN   
702       purine5566 (漂亮56)            Taipei   
703       purine5566 (漂亮56)            Taipei   
704        wuyiulin (機掰客家人)          Taichung   
705        wuyiulin (機掰客家人)            Taipei   
706         Giornno (快樂吳郭魚)            Taipei   
707         Giornno (快樂吳郭魚)            Taipei   
708              xxxqma (昀)            Taipei   
709              HMTking ()            Taipei   
710              HMTking ()            Taipei   
711            IkariD (怡卡滴)            Taipei   
712                     NaN               NaN   
713      spirit119 (一二三木頭人)            Taipei   
714            meimeiamy ()            Taipei   
715   Ccsleep (notsleepyet)            Taipei   
716            a435007 (陣雨)            Taipei   
717            sam45 (流水年華)          Taichung   
718               star07 ()          Taichung   
719   open20023 (open20023)            Taipei   
720   air121 (I GOT SPIRIT)            Taipei   
721        musashi4735 (武藏)            Taipei   
722           king1010 (玩酷)            Taipei   

                                               Content          Date_format  \
0                                                  NaN  2019-05-08 23:48:47   
1    發震時間：2019/05/08 23:48:47\n位置：北緯 24.81度，東經 122....  2019-05-08 23:49:28   
2    為什麼每次地震完\n\n小弟想上ptt 都上不去\n\n大家手腳都好快....馬上就上ptt...  2019-05-08 23:56:55   
3    我國東北的三寶是什麼？\n\n人參，貂皮，烏拉草\n\n還是\n\n人參，貂皮，鹿茸\n\n...  2019-05-09 00:43:35   
4                                                  NaN  2019-05-10 21:49:30   
5    幹  板橋有吧\n\n7777777777777777   快去開門阿  塊陶阿   補個字...  2019-05-10 21:50:07   
6    ※ 八卦板務請到 GossipPicket 檢舉板詢問\n※ a.張貼問卦請注意，充實文章內...  2019-05-11 08:50:16   
7                                                  NaN  2019-05-13 16:55:21   
8    幹你娘我在6樓快暈死了\n\n最近沒有下雨怎麼會地震\n\n辦公室好多人手機發出警報\n\n...  2019-05-13 16:55:49   
9    !!!\n\n剛剛開始搖\n\n在台北市有三級地震\n\n不知道大家有沒有感覺\n\n台北市...  2019-05-13 16:55:51   
10   地震文這麼久了\n怎麼還活著\n不曉得這個問卦會不會有人看到\n不要水桶我啦\n難得浮起來耶...  2019-05-13 16:55:54   
11   ※註:有電視或媒體有報導者，請勿使用爆卦!\n無重大八卦請

### 計算時間差

In [27]:
earthquake_index = []
for i in range(len(df_new)):
    if df_new.iloc[i]["IsOfficialReport"]==1:
        earthquake_index.append(i)
len(earthquake_index)

91

In [28]:
print(earthquake_index, sep = " ")

[0, 4, 7, 35, 44, 46, 52, 56, 64, 73, 75, 79, 81, 85, 91, 93, 96, 99, 104, 107, 110, 112, 143, 166, 180, 182, 184, 186, 191, 195, 199, 202, 204, 212, 234, 237, 240, 247, 250, 255, 260, 265, 268, 270, 276, 278, 280, 288, 293, 297, 299, 312, 318, 322, 324, 332, 338, 345, 357, 364, 397, 411, 440, 469, 471, 481, 488, 490, 493, 499, 505, 510, 515, 526, 530, 533, 540, 553, 556, 559, 561, 563, 577, 584, 593, 600, 604, 607, 654, 663, 712]


#### 試玩時間差

In [29]:
import datetime
import time

In [30]:
# Playground of datetime

# 地震報告的時間
dstr = df_new.iloc[0]["Date_format"]
d1 = datetime.datetime.strptime(dstr, "%Y-%m-%d %H:%M:%S")

# 地震爆文的時間
dstr = df_new.iloc[1]["Date_format"]
d2 = datetime.datetime.strptime(dstr, "%Y-%m-%d %H:%M:%S")



In [31]:
c = d2-d1
datetime.timedelta(0, 8, 562000)
#divmod(c.seconds, 60)
c.total_seconds()
#divmod(c.total_seconds(), 60)
#divmod(c.total_seconds(), 60)[1]

41.0

#### 實作時間差

In [32]:
# Set a list that contains 地震文與地震報告的時間差
TimeDifference = []
for j in range(len(df_new)):
    if df_new.iloc[j]["IsOfficialReport"] == 1: #如果是地震報告，則時間差為0
        dstr = df_new.iloc[j]["Date_format"] #找到時間標記
        d1 = datetime.datetime.strptime(dstr, "%Y-%m-%d %H:%M:%S") #Start Time 化為電腦的格式
        TimeDifference.append(0)
    else:
        dstr = df_new.iloc[j]["Date_format"] #找到時間標記
        d2 = datetime.datetime.strptime(dstr, "%Y-%m-%d %H:%M:%S") #地震文的發文時間化為電腦的格式
        d = d2-d1
        TimeDifference.append(d.total_seconds())

In [33]:
len(TimeDifference)

723

In [34]:
df_new["TimeDifference"] = TimeDifference

/Applications/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [35]:
len(df_new[df_new["TimeDifference"]<600]) #先以十分鐘為基準：超過十分鐘才發文的都不算想搶地震文

462

### 其他地震失敗文與地震爆文的時間差

# 計算距離

台灣地區若以 *北緯24度 東經121度* 即南投埔里附近為原點

則：

* 往東往西差一經度約差101公里

* 往南往北差一經度約差110公里

In [36]:
# Set a list that contains 地震文與地震報告的距離差
Distance = []
for j in range(len(df_new)):
    if df_new.iloc[j]["IsOfficialReport"] == 1: #如果是地震報告，則距離差為0
        X0 = df_new.iloc[j]["Longitude"] #找到經緯度
        Y0 = df_new.iloc[j]["Latitude"]  #當作原點
        Distance.append(0)
    else:
        X = df_new.iloc[j]["Longitude"] #找到經緯度
        Y = df_new.iloc[j]["Latitude"]
        d = np.sqrt( (101*(X-X0))**2 + (110*(Y-Y0))**2 )
        Distance.append(d)

In [37]:
df_new["Distance"] = Distance

/Applications/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [38]:
df_new[df_new["TimeDifference"]<600]

Author            City  \
0                          NaN             NaN   
1          DevNull (/dev/null)   Mountain View   
2            jackyuki1 (喵喵快來 )          Taipei   
4                          NaN             NaN   
5          tony77998 (偷尼柒柒玖玖捌)          Taipei   
7                          NaN             NaN   
8                     ros (螺絲)          Taipei   
9         samok (記者只會抄公社低卡記錄7)          Taipei   
10      world210 (KillerQueen)          Taipei   
11       LiaMeow (有時候甜蜜的心花朵朵開)          Taipei   
12                    a382 (斐)          Taipei   
13               cookie0215 ()          Taipei   
14       gcobs0834 (gcobs0834)          Taipei   
15          freelysoul (帥氣小活寶)          Taipei   
16      twin1949tw (台北Station)          Taipei   
17                 qx5 (學豬裝可愛)         Hsinchu   
18               harry68a (結衣)          Taipei   
19            a055035 (tomato)        Taichung   
20           longhandsome (阿駥)             NaN   
21          HayamaKami (台大彭于晏)          Taipei   
22                 JEWH (小田切腹)        Taichung   
23            airwalk66996 (平)          Taipei   
24             kevin600339 ( )          Taipei   
25               a386036 (骷髏怪)          Taipei   
26          chunyu0519 (廣東粥杰倫)       Chang-hua   
27                   ddt258 ()          Taipei   
28           jasontzymann (梵天)          Taipei   
29                 IHVH (讓愛滋生)     Nantou City   
30                a380919 (頭哥)          Taipei   
31               zalora (lara)  Kaohsiung City   
..                         ...             ...   
683         yjo566 (james8850)          Taipei   
684            stanleyup (5up)          Taipei   
685              love4654 (鍵盤)          Taipei   
686          ruby00514 (黑夜的星空)          Taipei   
687        rickphyman42 (理科廢男)          Taipei   
688             batutu (安樂岡花火)        Taichung   
689  rock04rock (CarlosChiang)    Taoyuan City   
690                  ET01 (ET)          Taipei   
691         a331432534 (jimmy)         Hsinchu   
692        ButterflyBao (一面湖水)          Taipei   
693        ButterflyBao (一面湖水)          Taipei   
694            aa9510883 (阿狄男)          Taipei   
695            aa9510883 (阿狄男)          Taipei   
696         e1608 (silentness)           Tampa   
697         e1608 (silentness)          Taipei   
698                 smd1201 ()          Taipei   
699                 smd1201 ()             NaN   
700           yutapon (川島  永嗣)          Taipei   
701           yutapon (川島  永嗣)             NaN   
712                        NaN             NaN   
713         spirit119 (一二三木頭人)          Taipei   
714               meimeiamy ()          Taipei   
715      Ccsleep (notsleepyet)          Taipei   
716               a435007 (陣雨)          Taipei   
717               sam45 (流水年華)        Taichung   
718                  star07 ()        Taichung   
719      open20023 (open20023)          Taipei   
720      air121 (I GOT SPIRIT)          Taipei   
721           musashi4735 (武藏)          Taipei   
722              king1010 (玩酷)          Taipei   

                                               Content          Date_format  \
0                                                  NaN  2019-05-08 23:48:47   
1    發震時間：2019/05/08 23:48:47\n位置：北緯 24.81度，東經 122....  2019-05-08 23:49:28   
2    為什麼每次地震完\n\n小弟想上ptt 都上不去\n\n大家手腳都好快....馬上就上ptt...  2019-05-08 23:56:55   
4                                                  NaN  2019-05-10 21:49:30   
5    幹  板橋有吧\n\n7777777777777777   快去開門阿  塊陶阿   補個字...  2019-05-10 21:50:07   
7                                                  NaN  2019-05-13 16:55:21   
8    幹你娘我在6樓快暈死了\n\n最近沒有下雨怎麼會地震\n\n辦公室好多人手機發出警報\n\n...  2019-05-13 16:55:49   
9    !!!\n\n剛剛開始搖\n\n在台北市有三級地震\n\n不知道大家有沒有感覺\n\n台北市...  2019-05-13 16:55:51   
10   地震文這麼久了\n怎麼還活著\n不曉得這個問卦會不會有人看到\n不要水桶我啦\n難得浮起來耶...  2019-05-13 16:55:54   
11   ※註:有電視或媒體有報導者，請勿使用爆卦!\n無重大八卦請勿使用此分類，否則視同濫用爆卦鬧板...  2019-05-13 16:55:59   
12   塊陶阿\n\n※註:有電視或媒體有報導者，請勿使用爆卦!\n無重大八卦請勿使用此分類，否則視...

# 簡化

In [39]:
# 簡化 只看那些時間差在十分鐘內的
df = df_new[df_new["TimeDifference"]<600]

In [40]:
# 拿掉official report，因為它們只是參照點，並非觀察值
df = df[df["IsOfficialReport"]==0]

# 新增三個columns

1. 發文的縣市是不是震央所在地

2. 發文的時間是不是深夜或凌晨

3. 所位在的地方的人口密度

In [41]:
df.columns

Index(['Author', 'City', 'Content', 'Date_format', 'Full IP', 'IP',
       'IsOfficialReport', 'Latitude', 'Longitude', 'Mobile', 'Nation', 'Push',
       'Title', 'YouWin', '位置', '深度', '編號', '規模', '震度', 'CenterLoc',
       'ShakingExtent', 'TimeDifference', 'Distance'],
      dtype='object')

In [42]:
set(df_new["位置"].dropna())

{'南投縣政府南偏東方  29.0  公里 (位於南投縣竹山鎮)(南投縣竹山鎮)',
 '南投縣政府南偏東方  29.2  公里 (位於南投縣竹山鎮)(南投縣竹山鎮)',
 '南投縣政府東偏北方  33.2  公里 (位於南投縣埔里鎮)(南投縣埔里鎮)',
 '嘉義縣政府南偏西方  9.7  公里 (位於嘉義縣鹿草鄉)(嘉義縣鹿草鄉)',
 '嘉義縣政府東偏北方  44.9  公里 (位於嘉義縣梅山鄉)(嘉義縣梅山鄉)',
 '嘉義縣政府東偏南方  22.7  公里 (位於嘉義縣中埔鄉)(嘉義縣中埔鄉)',
 '嘉義縣政府東偏南方  8.9  公里 (位於嘉義縣水上鄉)(嘉義縣水上鄉)',
 '基隆市政府東方  44.7  公里 (位於臺灣東北部海域)(臺灣東北部海域)',
 '宜蘭縣政府南偏東方  31.1  公里 (位於宜蘭縣近海)(宜蘭縣近海)',
 '宜蘭縣政府南偏東方  35.9  公里 (位於臺灣東部海域)(臺灣東部海域)',
 '宜蘭縣政府南偏東方  38.1  公里 (位於臺灣東部海域)(臺灣東部海域)',
 '宜蘭縣政府南偏東方  43.0  公里 (位於臺灣東部海域)(臺灣東部海域)',
 '宜蘭縣政府南偏東方  46.3  公里 (位於臺灣東部海域)(臺灣東部海域)',
 '宜蘭縣政府南偏西方  23.4  公里 (位於宜蘭縣南澳鄉)(宜蘭縣南澳鄉)',
 '宜蘭縣政府南偏西方  28.7  公里 (位於宜蘭縣南澳鄉)(宜蘭縣南澳鄉)',
 '宜蘭縣政府南方  44.6  公里 (位於宜蘭縣南澳鄉)(宜蘭縣南澳鄉)',
 '宜蘭縣政府東偏北方  14.1  公里 (位於宜蘭縣近海)(宜蘭縣近海)',
 '宜蘭縣政府東偏北方  14.5  公里 (位於宜蘭縣近海)(宜蘭縣近海)',
 '宜蘭縣政府東偏北方  30.3  公里 (位於臺灣東部海域)(臺灣東部海域)',
 '宜蘭縣政府東偏北方  32.3  公里 (位於臺灣東部海域)(臺灣東部海域)',
 '宜蘭縣政府東偏北方  8.4  公里 (位於宜蘭縣近海)(宜蘭縣近海)',
 '宜蘭縣政府東偏南方  24.9  公里 (位於臺灣東部海域)(臺灣東部海域)',
 '宜蘭縣政府東偏南方  44.4  公里 (位於臺灣東部海域)(臺灣東部海域)',
 '宜蘭縣政府東偏南方  69.6  公里 (位

In [43]:
set(df["City"].dropna())

{'Ashburn',
 'Chang-hua',
 'Chiayi City',
 'Douliu',
 'Hsinchu',
 'Hualien City',
 'Kaohsiung City',
 'Magong',
 'Miaoli',
 'Mountain View',
 'Nantou City',
 'Parsippany',
 'Petaling Jaya',
 'Pingtung City',
 'Taichung',
 'Tainan City',
 'Taipei',
 'Tampa',
 'Taoyuan City',
 'Taoyuan District',
 'Yilan',
 'Zhongli District',
 'Zhubei',
 '經度：121.000\u3000緯度：23.500'}

In [44]:
# 建字典
District_dict = {
 'Ashburn'       : "Ashburn, Virginia, U.S.A.",
 'Chang-hua'     : "彰化",
 'Chiayi City'   : "嘉義",
 'Douliu'        : "斗六",
 'Hsinchu'       : "新竹",
 'Hualien City'  : "花蓮",
 'Kaohsiung City': "高雄",
 'Magong'        : "馬公",
 'Miaoli'        : "苗栗",
 'Nantou City'   : "南投",
 'Parsippany'    : "Parsippany-Troy Hills, New Jersey, U.S.A.",
 'Petaling Jaya' : "Petaling Jaya, Petaling District, Selangor, Malaysia", 
 'Pingtung City' : "屏東",
 'Taichung'      : "台中",
 'Tainan City'   : "台南",
 'Taipei'        : "臺北",
 'Tampa'         : "Tampa, Florida, U.S.A",
 'Taoyuan City'  : "桃園",
 'Taoyuan District' : "桃園",
 'Yilan'         : "宜蘭",
 'Zhongli District' :"中壢",
 'Zhubei'        : "竹北"
}

In [45]:
City_Translated = []
for district in df["City"]:
    if district in District_dict:
        City_Translated.append(District_dict[district])
    else:
        City_Translated.append("NaN")

In [46]:
df["City_Translated"] = City_Translated

In [47]:
type(df.iloc[1]["位置"])

float

In [48]:
df.iloc[4]["位置"]

nan

In [49]:
InCenter = []
for i in range(len(df)):
    if str(df.iloc[i]["City_Translated"]) in str(df.iloc[i]["CenterLoc"]):
        InCenter.append(1)
    else:
        InCenter.append(0)
print(InCenter)
print(len(InCenter))

[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 

In [50]:
df["InCenter"] = InCenter

In [51]:
df[df["InCenter"]==1]

Author            City  \
58        HAHAcomet (值得信任的彗星小天使)  Kaohsiung City   
76           cake51640 (uuuu4ni)     Chiayi City   
111        online135 (98分美元宇宙星塵)    Hualien City   
113      openbestbook (吃蘿蔔不吐蘿蔔皮)          Taipei   
114  imjungyi923 (邊緣BL寫手 小旋風伊咪達)          Taipei   
115          Transfat (Transfat)          Taipei   
116         zxc37298312 (張瑋航的老二)          Taipei   
117                Random9 (優鮮沛)          Taipei   
119          ohohohya (安安你好我草泥馬)          Taipei   
120                  ko373328 ()          Taipei   
121            yangshuwei (yang)          Taipei   
122              youaregeek (吉克)          Taipei   
124              Huoyichung (路易)          Taipei   
125        windtrace (windtrace)          Taipei   
126               chuntai (努力打拼)          Taipei   
127              bossum (bossum)          Taipei   
129        FishJagor (FishJagor)          Taipei   
133            kaky (菩提本無樹明鏡亦非台)          Taipei   
134      ray1990715 (Wizard.geo)          Taipei   
135         qq014 (TreeThousand)          Taipei   
192            yoyodiy (廢文心得文大師)          Miaoli   
196                iasyt (iasyt)  Kaohsiung City   
238               minagoroshi ()           Yilan   
241            Tiandai (Tiandai)           Yilan   
273             keybo (唱歌給我聽(b))  Kaohsiung City   
279                leesrnat (大緑)           Yilan   
298               minagoroshi ()           Yilan   
333             henrygjs22 (錢庭寶)          Taipei   
339            ga86041 (ga86041)    Hualien City   
414              x21198 (地殼的傳令兵)           Yilan   
445          oscar2022 (對不起，我愛你)    Hualien City   
484            LisaPrincess (莎莎)     Nantou City   
489                    astrayzip    Hualien City   
500                    astrayzip    Hualien City   
507                jt910203 (aa)    Hualien City   
516          bomb1000 (bomb1000)          Taipei   
554                    astrayzip    Hualien City   
560            Whitetext (0 . 0)    Hualien City   
594                    astrayzip    Hualien City   
601                    astrayzip    Hualien City   
608                   B0Y0 (假柏濕)           Yilan   
609                   B0Y0 (假柏濕)           Yilan   
615             pt11235813 (PTC)           Yilan   
655   AntitheApple (蘋果? 那是什麼好吃嗎)    Hualien City   
656   AntitheApple (蘋果? 那是什麼好吃嗎)    Hualien City   
664            Tiandai (Tiandai)           Yilan   

                                               Content          Date_format  \
58   \n\n\n\n剛剛地震發生前 八卦人數大概是14XXX\n\n地震後的現在 馬上增加到19...  2019-05-31 23:59:26   
76                                             幹地震\n\n  2019-06-26 05:09:08   
111  地震!!!\n\n\n※註:有電視或媒體有報導者，請勿使用爆卦!\n無重大八卦請勿使用此分類...  2019-08-06 09:20:36   
113  \n哞哞哞哞哞哞\n\n小精靈縮4台北市附近發生有感地震  哞~~\n五樓快承認翻身跌下床\...  2019-08-07 01:01:01   
114                   新北中和有感\n\n雖然只小晃一下\n\n還以為是我頭暈\n\n  2019-08-07 01:01:11   
115                                             地震\n\n  2019-08-07 01:01:12   
116                                           台北有感\n\n  2019-08-07 01:01:18   
117                                             地震\n\n  2019-08-07 01:01:19   
119                           啊啊啊啊啊啊\n\n新北\n\n很大很晃\n\n  2019-08-07 01:01:27   
120                               \n\n震震震震\n\n跑跑跑跑\n\n  2019-08-07 01:01:38   
121                                           振振有詞\n\n  2019-08-07 01:01:41   
122                                            塊陶R\n\n  2019-08-07 01:01:47   
124                              台北超大 好可怕rrrrrrrrr\n\n  2019-08-07 01:01:58   
125  搖好大...\n\n怕,jpg\n\n\n※註:有電視或媒體有報導者，請勿使用爆卦!\n無重...  2019-08-07 01:02:21   
126                               靠杯台北好大\n\n大家都還好嗎\n\n  2019-08-07 01:02:59   
127                                     \n地震啊！！快逃啊\n\n  2019-08-07 01:03:06   
129                                    天啊啊啊啊 地震好可怕\n\n  2019-08-07 01:03:48   
133  https://www.cwb.gov.tw/V8/C/E/EQ/EQ108000-0807...  2019-08-07 01:09:27   
134  \n震央在北投!!!\n\n\n                   中央氣象局地震測報中心...  2019-08-07 01:10:06   
1

### 尖峰與離峰時段：另外兩個dummies

尖峰時段就設定在peak的前後兩個小時 

* 22:00~2:00

* 13:00~17:00

離峰時間設定在

* 4:00~8:00

* 18:00~20:00

In [52]:
df.iloc[1]["Date_format"]

'2019-05-08 23:56:55'

In [53]:
dstr = df.iloc[32]["Date_format"]
d1 = datetime.datetime.strptime(dstr, "%Y-%m-%d %H:%M:%S")
d1.timetuple()
d1.timetuple()[3] #表示23時

0

In [54]:
# Set a list that contains dummies for peak & offpeak
IsPeak = []
OffPeak = []
for j in range(len(df)):
    dstr = df.iloc[j]["Date_format"] #找到時間標記
    d1 = datetime.datetime.strptime(dstr, "%Y-%m-%d %H:%M:%S") #Start Time 化為電腦的格式
    hour = d1.timetuple()[3] #拿到使用者是在幾點發文的
    
    if hour in [22, 23, 0, 1, 13, 14, 15, 16]:
        IsPeak.append(1)
    else:
        IsPeak.append(0)
    
    if hour in [4, 5, 6, 7, 18, 19]:
        OffPeak.append(1)
    else:
        OffPeak.append(0)
print(len(IsPeak), len(OffPeak))

371 371


In [55]:
df["IsPeak"] = IsPeak
df["OffPeak"] = OffPeak

In [56]:
df.head()

Author           City  \
1   DevNull (/dev/null)  Mountain View   
2     jackyuki1 (喵喵快來 )         Taipei   
5   tony77998 (偷尼柒柒玖玖捌)         Taipei   
8              ros (螺絲)         Taipei   
9  samok (記者只會抄公社低卡記錄7)         Taipei   

                                             Content          Date_format  \
1  發震時間：2019/05/08 23:48:47\n位置：北緯 24.81度，東經 122....  2019-05-08 23:49:28   
2  為什麼每次地震完\n\n小弟想上ptt 都上不去\n\n大家手腳都好快....馬上就上ptt...  2019-05-08 23:56:55   
5  幹  板橋有吧\n\n7777777777777777   快去開門阿  塊陶阿   補個字...  2019-05-10 21:50:07   
8  幹你娘我在6樓快暈死了\n\n最近沒有下雨怎麼會地震\n\n辦公室好多人手機發出警報\n\n...  2019-05-13 16:55:49   
9  !!!\n\n剛剛開始搖\n\n在台北市有三級地震\n\n不知道大家有沒有感覺\n\n台北市...  2019-05-13 16:55:51   

           Full IP              IP  IsOfficialReport  Latitude  Longitude  \
1   35.236.191.107  35.236.191.107               0.0    37.404   -122.075   
2   112.104.89.169  112.104.89.169               0.0    25.048    121.532   
5    58.115.130.21   58.115.130.21               0.0    25.048    121.532   
8    36.231.45.115   36.231.45.115               0.0    25.048    121.532   
9   163.21.235.221  163.21.235.221               0.0    25.048    121.532   

                     Mobile   ...    規模  震度  \
1              Google Cloud   ...   NaN NaN   
2            Digital United   ...   NaN NaN   
5  Hoshin Multimedia Center   ...   NaN NaN   
8                     HiNet   ...   NaN NaN   
9   Taiwan Academic Network   ...   NaN NaN   

                              CenterLoc  ShakingExtent TimeDifference  \
1  宜蘭縣政府東方  74.2  公里 (位於臺灣東部海域)(臺灣東部海域)            2.0           41.0   
2  宜蘭縣政府東方  74.2  公里 (位於臺灣東部海域)(臺灣東部海域)            2.0          488.0   
5   宜蘭縣政府東南方  30.3  公里 (位於宜蘭縣近海)(宜蘭縣近海)            3.0           37.0   
8    宜蘭縣政府東方  14.1  公里 (位於宜蘭縣近海)(宜蘭縣近海)            3.0           28.0   
9    宜蘭縣政府東方  14.1  公里 (位於宜蘭縣近海)(宜蘭縣近海)            3.0           30.0   

       Distance City_Translated  InCenter  IsPeak OffPeak  
1  24739.820786             NaN         0       1       0  
2     99.955575              臺北         0       1       0  
5     71.898385              臺北         0       0       0  
8     49.557932              臺北         0       1       0  
9     49.557932              臺北         0       1       0  

[5 rows x 27 columns]

## 人口密度

利用臺灣行政區面積表及臺灣行政區人口列表的資料計算

https://zh.wikipedia.org/wiki/臺灣行政區人口密度表

In [57]:
# 建字典 人口密度 單位：人/每平方公里
popDensity_dict = {
    "彰化" : 1182.02,
    "嘉義" : 263.38,
    "斗六" : 526.38, #使用縣市層級，i.e.雲林縣的人口密度
    "新竹" : 4318.72,
    "花蓮" : 70.32,
    "高雄" : 939.35,
    "馬公" : 827.83, #使用澎湖縣
    "苗栗" : 299.04,
    "南投" : 120.03,
    "屏東" : 294.05,
    "台中" : 1271.61,
    "台南" : 857,
    "臺北" : 9662.94,
    "桃園" : 1848.28,
    "宜蘭" : 211.78,
    "中壢" : 1848.28,#使用桃園
    "竹北" : 4318.72 #使用新竹
}

In [58]:
PopDensity = []
for city in df["City_Translated"]:
    if city in popDensity_dict.keys():
        PopDensity.append(popDensity_dict[city])
    else:
        PopDensity.append(float("NaN"))
len(PopDensity)

371

In [59]:
df["PopDensity"] = PopDensity

In [60]:
df.head()

Author           City  \
1   DevNull (/dev/null)  Mountain View   
2     jackyuki1 (喵喵快來 )         Taipei   
5   tony77998 (偷尼柒柒玖玖捌)         Taipei   
8              ros (螺絲)         Taipei   
9  samok (記者只會抄公社低卡記錄7)         Taipei   

                                             Content          Date_format  \
1  發震時間：2019/05/08 23:48:47\n位置：北緯 24.81度，東經 122....  2019-05-08 23:49:28   
2  為什麼每次地震完\n\n小弟想上ptt 都上不去\n\n大家手腳都好快....馬上就上ptt...  2019-05-08 23:56:55   
5  幹  板橋有吧\n\n7777777777777777   快去開門阿  塊陶阿   補個字...  2019-05-10 21:50:07   
8  幹你娘我在6樓快暈死了\n\n最近沒有下雨怎麼會地震\n\n辦公室好多人手機發出警報\n\n...  2019-05-13 16:55:49   
9  !!!\n\n剛剛開始搖\n\n在台北市有三級地震\n\n不知道大家有沒有感覺\n\n台北市...  2019-05-13 16:55:51   

           Full IP              IP  IsOfficialReport  Latitude  Longitude  \
1   35.236.191.107  35.236.191.107               0.0    37.404   -122.075   
2   112.104.89.169  112.104.89.169               0.0    25.048    121.532   
5    58.115.130.21   58.115.130.21               0.0    25.048    121.532   
8    36.231.45.115   36.231.45.115               0.0    25.048    121.532   
9   163.21.235.221  163.21.235.221               0.0    25.048    121.532   

                     Mobile    ...      震度  \
1              Google Cloud    ...     NaN   
2            Digital United    ...     NaN   
5  Hoshin Multimedia Center    ...     NaN   
8                     HiNet    ...     NaN   
9   Taiwan Academic Network    ...     NaN   

                              CenterLoc ShakingExtent  TimeDifference  \
1  宜蘭縣政府東方  74.2  公里 (位於臺灣東部海域)(臺灣東部海域)           2.0            41.0   
2  宜蘭縣政府東方  74.2  公里 (位於臺灣東部海域)(臺灣東部海域)           2.0           488.0   
5   宜蘭縣政府東南方  30.3  公里 (位於宜蘭縣近海)(宜蘭縣近海)           3.0            37.0   
8    宜蘭縣政府東方  14.1  公里 (位於宜蘭縣近海)(宜蘭縣近海)           3.0            28.0   
9    宜蘭縣政府東方  14.1  公里 (位於宜蘭縣近海)(宜蘭縣近海)           3.0            30.0   

       Distance  City_Translated InCenter  IsPeak  OffPeak PopDensity  
1  24739.820786              NaN        0       1        0        NaN  
2     99.955575               臺北        0       1        0    9662.94  
5     71.898385               臺北        0       0        0    9662.94  
8     49.557932               臺北        0       1        0    9662.94  
9     49.557932               臺北        0       1        0    9662.94  

[5 rows x 28 columns]

# 看一些簡單的統計量

In [61]:
np.mean(df["TimeDifference"]) #平均發文間隔

127.38814016172506

In [62]:
np.mean(df[df["YouWin"]==1]["TimeDifference"]) #地震爆文平均發文間隔

46.55140186915888

In [63]:
#df["YouWin"]
np.mean(df["YouWin"]) #搶到爆文的比率

0.2884097035040431

In [64]:
np.mean(df["Distance"]) #平均距離

451.6240886074431

In [65]:
np.mean(df[df["YouWin"]==1]["Distance"]) #那些搶到爆文的人平均而言離震央的距離

842.461655696156

In [66]:
# 位處震央的人的比例
np.mean(df["InCenter"])

0.12398921832884097

In [67]:
np.mean(df["IsPeak"]) #有36%的人在尖峰時段上ptt

0.3692722371967655

In [68]:
np.mean(df["OffPeak"]) #有36%的人在尖峰時段上ptt

0.3261455525606469

In [69]:
np.mean(df[df["IsPeak"]==1]["YouWin"]) #在尖峰時段搶到爆文的比例

0.3284671532846715

In [70]:
np.mean(df[df["OffPeak"]==1]["YouWin"]) #在離峰時段搶到爆文的比例

0.1652892561983471

In [71]:
np.mean(df[df["YouWin"]==1]["PopDensity"]) #那些搶到爆文的人所在的城市平均而言的人口密度

4616.575212765959

# 拿掉不要的columns以到R跑迴歸

In [72]:
df.columns

Index(['Author', 'City', 'Content', 'Date_format', 'Full IP', 'IP',
       'IsOfficialReport', 'Latitude', 'Longitude', 'Mobile', 'Nation', 'Push',
       'Title', 'YouWin', '位置', '深度', '編號', '規模', '震度', 'CenterLoc',
       'ShakingExtent', 'TimeDifference', 'Distance', 'City_Translated',
       'InCenter', 'IsPeak', 'OffPeak', 'PopDensity'],
      dtype='object')

### 變數說明

'Date_format' : 時間

'YouWin' : 是否是地震爆文

'TimeDifference': 發文時間與中央氣象局的地震報告標注的地震發生時間的差

'Distance': IP位址反查得到的經緯度與地震震央經緯度的距離

'ShakingExtent':震度

'InCenter':是否為在震央所在縣市

'City_Translated':發文者所在位置

'CenterLoc':震央

'IsPeak' : 是否在尖峰時段發文

'OffPeak': 是否在離峰時段發文

'PopDensity' : 發文者所在的縣市的人口密度



'Author':作者ID

'Content':發文內容

'Mobile':使用的電信業者

'Push':推文內容

'Title':發文標題

In [73]:
df_R = df[['Date_format','YouWin','TimeDifference', 'Distance', 'ShakingExtent',
           'InCenter', 'City_Translated', 'CenterLoc', 'IsPeak', 'OffPeak', 'PopDensity',
           'Title', 'Author', 'Content', 'Mobile', 'Push']]

# Save file

In [116]:
df_R.to_csv("combined_forR.csv", encoding="utf-8")

# 畫地圖的資料

In [76]:
df_new.columns

Index(['Author', 'City', 'Content', 'Date_format', 'Full IP', 'IP',
       'IsOfficialReport', 'Latitude', 'Longitude', 'Mobile', 'Nation', 'Push',
       'Title', 'YouWin', '位置', '深度', '編號', '規模', '震度', 'CenterLoc',
       'ShakingExtent', 'TimeDifference', 'Distance'],
      dtype='object')

In [79]:
df_graph = df_new[["Date_format", "IsOfficialReport", "Latitude", "Longitude", "震度", "YouWin"]]

In [80]:
df_graph.head()

Date_format  IsOfficialReport  Latitude  Longitude   震度  YouWin
0  2019-05-08 23:48:47               1.0    24.820    122.490  2.0     NaN
1  2019-05-08 23:49:28               0.0    37.404   -122.075  NaN     1.0
2  2019-05-08 23:56:55               0.0    25.048    121.532  NaN     0.0
3  2019-05-09 00:43:35               0.0    25.048    121.532  NaN     0.0
4  2019-05-10 21:49:30               1.0    24.500    121.920  3.0     NaN

In [81]:
df_graph.to_csv("Data_graphing.csv", encoding="utf-8")